### PR TITLE
[FIX] website_slides: prevent redirection issues

### PR DESCRIPTION
**Pull request by aath-odoo**
**Description of the issue/feature this PR addresses:**
This PR address the issue of redirecting to courses page with /invite_error when trying to update the optimize seo.

**Current behavior before PR:**
Previously, when editing Optimize SEO and clicking "Save," it redirected
to /invite_error in Courses instead of remaining on the same URL after saving, which was caused due to not receiving seoName.

**Desired behavior after PR is merged:**
Now after saving, we stay on the same course URL without being redirected
to /invite_error. This fix includes checking if the seoContext doesn't have a seoName , in which case we redirect to seoNameDefault to prevent redirection to /invite_error on the Courses page.

**Task**-3688285

### CODE REVIEW

File: addons/website/static/src/components/dialog/seo.js [hunk 1]
```diff
@@ -457,6 +457,9 @@ export class OptimizeSEODialog extends Component {
             if (seoContext.seoName !== this.previousSeoName) {
                 data.seo_name = seoContext.seoName;
             }
+            if (!seoContext.seoName) {
+                seoContext.seoName = this.seoNameDefault;
```
Comment (lm4649) [inline addons/website/static/src/components/dialog/seo.js:461]: As this guard clause is here to prevent redirection issues, why not putting it just before the redirection (`this.website.goToWebsite`...)?
It seems that the 'if canEdit' blocks are here to update the data to pass to the orm whenever it's necessary. We may keep them limited to that purpose.

Just a suggestion:
```
        this.website.goToWebsite({
            path: this.url.replace(
                this.previousSeoName || this.seoNameDefault,
                seoContext.seoName || this.seoNameDefault
            ),
        });
```

**Review by lm4649 — COMMENTED** [review]
Hello @aath-odoo :wave: 

Nice catch :+1: 
Though I have a couple of remarks / suggestions before we move on:

The commit header, (the part right after ':') should include an infinitive verb, we should be able to complete the following sentence with: if applied this commit will...(e.g. : prevent redirection issues).

As for the commit message, a bit of explanation on how you fix the issue is greatly welcome :pray: 

And also, the fix is great, but could be better placed in the code, to keep its readability/meaning.

Thanks for the great work :+1:

File: addons/website/static/src/components/dialog/seo.js (hunk 1 above)
Comment (aath-odoo) [reply addons/website/static/src/components/dialog/seo.js]: Hello @lm4649, Thanks for pointing out, adapted the changes. Thanks :)

**Review by lm4649 — APPROVED** [review]
Hi @aath-odoo,

Thank you for this quick update. LGTM :+1: 
Not sure this redirection issue is restricted to the Courses section of the website. During the next steps, it's should be good to check the scope of the issue and update the commit message consequently :pray:

File: addons/website/static/src/components/dialog/seo.js [hunk 2]
```diff
@@ -465,7 +465,7 @@ export class OptimizeSEODialog extends Component {
                 'website_id': this.website.currentWebsite.id,
             },
         });
-        this.website.goToWebsite({path: this.url.replace(this.previousSeoName || this.seoNameDefault, seoContext.seoName)});
+        this.website.goToWebsite({path: this.url.replace(this.previousSeoName || this.seoNameDefault, seoContext.seoName || this.seoNameDefault)});
```
Comment (rdeodoo) [inline addons/website/static/src/components/dialog/seo.js:468]: The change should be in the `/slides` controller, which is the root cause, this is just one of the possibly multiple cases highlighting the bug in the `/slides` controller.

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Hello :wave: 

The question was [asked](https://discord.com/channels/678381219515465750/687337739129323543/1230550115731640350) on our discord, I will also copy paste it here for tracking purpose :+1: 

-----------------

There is no explanation whatsoever in the commit message about why is the error happening and neither why is this targetting 16.3, but after checking a bit, it's actually because of [this](https://github.com/odoo/odoo/commit/69277f846ea683ac2020f653256f9a25707e1a5c#diff-d5d755f5bf3898c934729c526a42b3b043e57737950cccc05c22643eef780b3dR479-R481), hardcoded integer as converter in the route instead of using modelconverter.
I don't know why it was done like that, did not check further yet, but even if it was wrong (again, did not check), we can't change that in stable, so a fix like the one of the PR might do the trick, but it's just fixing the consequence of the issue in this very particular flow (seo save), which is not ideal.

What I'd do, is adding some code inside the `/slide` route to not enter this condition and just suppose that if the id is negative, it just need to be converted to positive id:
```py
        # Check access rights
        if channel_id and not channel:
            channel = request.env['slide.channel'].browse(channel_id).exists()
            if not channel:
                return self._redirect_to_slides_main('no_channel')
```
I guess it's trying to browse `-2` which obviously fail.

But fixing it there will make the route robust to all caller, which would be better.
You can actually try with any controller, it will work: `/shop/-2` or `/blog/-2`
While it won't work with `/slide/-2`, but it should.

---------------------

**Question**:  what does "-2" represent exactly?

**Answer**: It's just a `slug` (~ the model converter), so when you reach `/slide/my-slide-3`, `my-slide-3` is the slug used by the modelconverter to extract the (orm) record from the route `/slides/<model("slide.channel"):channel>`
It's basically just a `<string>-<id>`.

Converters are part of native werkzeug "framework" (IntConverter etc, like `<int:channel_id>`), `ModelConverter` is just our own implementation of our own Converter with Odoo record/model.

And `-2` is the same as `string-2`, it's an empty string followed by a `-` separator and an `id`.

We don't really care about the `string` part, it can be whatever, you can try to go to `/slides/rderderde-1`, it will redirect to the correct url with the correct slugified name of the record slide 1. Only `-1` will be helpful to match.

But the custom code + the usage of IntConverter instead of ModelConverter in the slide route is breaking that

------------------------

tl;dr: don't change code in seo.js (consequence of the bug), but fix the /slides route (root cause of the bug) so it does not failed when called with `/slides/-2`

**Comment by aath-odoo** [comment]
@awa-odoo Did the changes in the route itself looks like it does the fix if we fetch the record and check the existence of the record could you take a look Thanks.

File: addons/website_slides/controllers/main.py [hunk 3]
```diff
@@ -526,8 +526,10 @@ def channel(self, channel=False, channel_id=False, category=None, category_id=Fa
 
         # Check access rights
         if channel_id and not channel:
-            channel = request.env['slide.channel'].browse(channel_id).exists()
```
Comment (awa-odoo) [inline addons/website_slides/controllers/main.py:529]: Would make it even more explicit, something like this:
```py
if channel_id < 0:
    # the string part of the channel "slugification" can be blank
    # meaning it can be "/slides/taking-care-of-trees-2" OR just "/slides/-2" if the first part is blank
    # as we use a IntConverter on the route definition, this will pick up a negative ID
    # (the IntConverter is necessary as we want a custom page in case the user can't access the course)
    channel_id = abs(channel_id)
```

File: addons/website_slides/tests/test_attendee.py [hunk 4]
```diff
@@ -441,10 +441,12 @@ def test_invite_route_errors_handling(self):
         self.channel.visibility = 'members'
 
         # No such channel
-        invite_url = "/slides/-1"
-        res = self.url_open(invite_url)
-        self.assertEqual(res.status_code, 200)
-        self.assertTrue('/slides?invite_error=no_channel' in res.url, "This channel does not exist. Redirect to the main /slides page.")
+        channel_id = -123
```
Comment (awa-odoo) [inline addons/website_slides/tests/test_attendee.py:444]: Would instead rework this test case to use "negative-id" on self.channel and make sure it works.

In addition, would pick up the max existing ID (search sorted by id desc limit 1), and add 1 to that number, to make sure it redirects to "no_channel" in case we have something that does not exist.

That way we test everything.

**Review by awa-odoo — COMMENTED** [review]
Hello @aath-odoo 

Moving in the right direction but I would like to make it more explicit, see underlying comments.

In addition, I would rework your commit message (which seems outdated) and add the explanations from my sample code comment.
That way we explain why it happens and not just how to reproduce it, which is more important.

Thanks & cheers!

File: addons/website_slides/controllers/main.py [hunk 5]
```diff
@@ -525,9 +525,18 @@ def channel(self, channel=False, channel_id=False, category=None, category_id=Fa
...
+            # (the IntConverter is necessary as we want a custom page in case the user can't access the course)
+            channel_id = abs(channel_id)
+
         if channel_id and not channel:
-            channel = request.env['slide.channel'].browse(channel_id).exists()
-            if not channel:
+            channel = request.env['slide.channel'].browse(channel_id)
+            if channel.exists():
+                channel_id = False
+            else:
```
[the remark this answers is not in the document]
Comment (aath-odoo) [reply addons/website_slides/controllers/main.py]: In scenarios where the channel_id is positive but does not correspond to an existing course, such as in the test case when attempting to increment the ID beyond the last known channel ID, the redirection should be to invite_error=no_channel.
So, we're validating if the channel exists with the channel_id or not.

[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/website_slides/controllers/main.py]: But the previous version of the code already did that, right?

[the remark this answers is not in the document]
Comment (aath-odoo) [reply addons/website_slides/controllers/main.py]: Ah, Okay yes that'll do. Thanks for pointing out

File: addons/website_slides/tests/test_attendee.py [hunk 6]
```diff
@@ -441,10 +441,15 @@ def test_invite_route_errors_handling(self):
         self.channel.visibility = 'members'
 
         # No such channel
```
Comment (awa-odoo) [inline addons/website_slides/tests/test_attendee.py:443]: Rename to `# slugification can give: "/slides/-ID" which should work, despite resulting in a negative ID`

File: addons/website_slides/tests/test_attendee.py [hunk 7]
```diff
@@ -441,10 +441,15 @@ def test_invite_route_errors_handling(self):
...
-        invite_url = "/slides/-1"
+        invite_url = f'/slides/-{self.channel.id}'
         res = self.url_open(invite_url)
         self.assertEqual(res.status_code, 200)
-        self.assertTrue('/slides?invite_error=no_channel' in res.url, "This channel does not exist. Redirect to the main /slides page.")
+        self.assertTrue(invite_url in res.url)
+        max_channel_id = self.env['slide.channel'].search([], order='id desc', limit=1).id
```
Comment (awa-odoo) [inline addons/website_slides/tests/test_attendee.py:448]: Add a blank line and then the python code comment mentioning "No such channel".

File: addons/website_slides/tests/test_attendee.py [hunk 8]
```diff
@@ -441,10 +441,15 @@ def test_invite_route_errors_handling(self):
...
+        invite_url = f'/slides/{max_channel_id + 1}'
+        res = self.url_open(invite_url)
+        self.assertEqual(res.status_code, 200)
+        self.assertTrue('/slides?invite_error=no_channel' in res.url, "Should have redirected to the 'no_channel' page as this channel ID does not exist")
```
Comment (awa-odoo) [inline addons/website_slides/tests/test_attendee.py:452]: Put the string explanation on its own code line to avoid a very long line.

File: addons/website_slides/controllers/main.py [hunk 9]
```diff
@@ -525,6 +525,13 @@ def channel(self, channel=False, channel_id=False, category=None, category_id=Fa
                 }
```
Comment (awa-odoo) [inline addons/website_slides/controllers/main.py:525]: Commit header guidelines: Change "OptimizeSEO" to "website_slides" as it is the module we are modifying.

File: addons/website_slides/controllers/main.py [hunk 10]
```diff
@@ -525,6 +525,13 @@ def channel(self, channel=False, channel_id=False, category=None, category_id=Fa
...
 
         # Check access rights
```
Comment (awa-odoo) [inline addons/website_slides/controllers/main.py:527]: This code comment should stay right before the next if.

**Review by awa-odoo — COMMENTED** [review]
@aath-odoo 

Some more polishing and we should be good.

Thanks & cheers!

**Comment by awa-odoo** [comment]
@aath-odoo Note that your test doesn't seem to work, runbot is red.

**Comment by aath-odoo** [comment]
> @aath-odoo Note that your test doesn't seem to work, runbot is red.

Seems like we needed authentication for read access should be fine now, Thanks :)

**Comment by awa-odoo** [comment]
@robodoo r+

@aath-odoo Thanks for the updates!
